src/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// Converts an object or value to or from JSON.
+    /// </summary>
+    /// <typeparam name="T">The <see cref="Type"/> to convert.</typeparam>
+    public abstract class JsonConverter<T> : JsonConverter
+    {
+        /// <summary>
+        /// When overidden, constructs a new <see cref="JsonConverter{T}"/> instance.
+        /// </summary>
+        protected internal JsonConverter() { }
+
+        /// <summary>
+        /// Determines whether the type can be converted.
+        /// </summary>
+        /// <remarks>
+        /// The default implementation is to return True when <paramref name="typeToConvert"/> equals typeof(T).
+        /// </remarks>
+        /// <param name="typeToConvert"></param>
+        /// <returns>True if the type can be converted, False otherwise.</returns>
+        public override bool CanConvert(Type typeToConvert)
+        {
+            return typeToConvert == typeof(T);
+        }
+
+        /// <summary>
+        /// Read and convert the JSON to T.
+        /// </summary>
+        /// <remarks>
+        /// The following exceptions are caught and re-thrown as <see cref="JsonException"/>:
+        /// <see cref="ArgumentException"/>
+        /// <see cref="FormatException"/>
+        /// <see cref="InvalidOperationException"/>
+        /// <see cref="OverflowException"/>
+        /// </remarks>
+        /// <param name="reader">The <see cref="Utf8JsonReader"/> to read from.</param>
+        /// <param name="typeToConvert">The <see cref="Type"/> being converted.</param>
+        /// <param name="options">The <see cref="JsonSerializerOptions"/> being used.</param>
+        /// <returns>The value that was converted.</returns>
+        public abstract T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options);
```

Comment:
Do we have any examples of when the input isn't all in the here? Since this is blocking what part of the stack does the buffering and yielding?

cc @rynowak @JamesNK

Reply:
The other question is how would it work if I need to call back into the de-serializer to do a sub-object? Aren't those APIs going to be async or are we also adding sync versions?

Reply:
> Do we have any examples of when the input isn't all in the here? Since this is blocking what part of the stack does the buffering and yielding?

Yes in [CustomConverterTests.ReadAhead.cs
](https://github.com/dotnet/corefx/pull/38578/files#diff-a5b0ff59e620fd1a4e20e45f39933358)

Basically if using a Stream + processing JSON object (StartObject token) + there is a converter for the given type, then we pre-read the current JSON depth to ensure the buffer will contain all of the information. This means that normal POCOs and arrays won't pre-read; the only case where we pre-read with built-in types is with `JsonElement`. cc @JeremyKuhne 

Also I'd like to write a new type of converter that won't have to pre-read but that won't be in 3.0; it would have a different programming model including "callbacks" per property and no access to the reader\writer.

> The other question is how would it work if I need to call back into the de-serializer to do a sub-object? Aren't those APIs going to be async or are we also adding sync versions?

See [CustomConverterTests.Callback.cs ](https://github.com/dotnet/corefx/pull/38578/files#diff-9b72df50ce20be464d9c19e4c13e20fc) for examples. Currently just sync is supported\necessary due to the pre-read mentioned above.

Reply:
> Also I'd like to write a new type of converter that won't have to pre-read but that won't be in 3.0; it would have a different programming model including "callbacks" per property and no access to the reader\writer.

I'd like to discuss that with you and as long as there's a path to this post 3.0 I'm ok. The current buffering simplifies writing converters but maybe there's a sane way to expose the ReadStack and the WriteStack in the converters as well.

Reply:
>  but maybe there's a sane way to expose the ReadStack and the WriteStack in the converters as well.

To enumerate some options:
1) Next-gen converter as mentioned above
2) Add an opaque "SerializationState" type that is passed into a convert's Read\Write method (like JsonConverter<T>) and passed back to the JsonSerializer.Parse\ToString\etc.
3) Re-use `JsonSerializationOptions`. Create a subclass of `JsonSerializationOptions` that holds a reference to the real options class. It has an internal "SerializationState" property. An instance of this is created for each call into a converter and is assumed to be stack-based (not held globally).
4) Similar to (3), but re-use reader\writer state.
5) Do nothing. Calls into the serializer lose the option settings, Path, etc.
6) ...

Reply:
Lets sync up offline on this. I'd request that we don't write our internal converters on the public API (unless we can verify it won't cause any harmful buffering/copying).